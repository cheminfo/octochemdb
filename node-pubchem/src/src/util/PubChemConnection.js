'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
