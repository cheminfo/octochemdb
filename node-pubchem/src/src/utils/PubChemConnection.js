import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
