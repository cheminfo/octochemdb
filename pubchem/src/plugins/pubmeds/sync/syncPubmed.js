import firstPubmedsImport from './firstPubmedsImport.js';
import incrementalPubmedsImport from './incrementalPubmedsImport.js';
/**
 * @description perform first and incremental import of pubmeds
 * @param {*} connection - mongo connection
 */
export async function sync(connection) {
  await firstPubmedsImport(connection);
  await incrementalPubmedsImport(connection);
}
