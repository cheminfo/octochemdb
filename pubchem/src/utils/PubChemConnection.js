import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
