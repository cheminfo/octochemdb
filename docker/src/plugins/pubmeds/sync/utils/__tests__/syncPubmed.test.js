import { PubChemConnection } from '../../../../../utils/PubChemConnection.js';
import { sync } from '../../syncPubmed';

jest.setTimeout(10000);
test('syncPubmed First Importation', async () => {
  const connection = new PubChemConnection();
  await sync(connection);
  const collection = await connection.getCollection('pubmeds');
  const collectionEntry = await collection.find({ _id: 14248047 }).limit(1);
  const result = await collectionEntry.next();
  expect(result).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
test('syncPubmed Incremental Importation', async () => {
  const connection = new PubChemConnection();
  const collection = await connection.getCollection('pubmeds');
  const collectionEntryIncremental = await collection
    .find({ _id: 17200418 })
    .limit(1);
  const resultIncremental = await collectionEntryIncremental.next();
  expect(resultIncremental).toMatchSnapshot();
  if (connection) {
    await connection.close();
  }
});
