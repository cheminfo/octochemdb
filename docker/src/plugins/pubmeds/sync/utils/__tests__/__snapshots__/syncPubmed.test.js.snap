// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncPubmed First Importation 1`] = `
{
  "_id": 14248047,
  "_seq": 2,
  "data": {
    "article": {
      "authors": [
        {
          "firstName": "G",
          "initials": "G",
          "lastName": "HEVESY",
        },
      ],
      "journal": {
        "iSSN": "0030-6002",
        "isoAbbreviation": "Orv Hetil",
        "pubDate": {
          "Day": 3,
          "Month": "Jan",
          "Year": 1965,
        },
        "title": "Orvosi hetilap",
      },
      "language": "hun",
      "publicationTypeList": [
        "Journal Article",
      ],
      "title": "USE OF ISOTOPE INDICATORS IN MEDICINE.",
    },
    "chemicals": [
      {
        "nameOfSubstance": "Isotopes",
      },
      {
        "nameOfSubstance": "Radioisotopes",
      },
    ],
    "dateCompleted": {
      "Day": 1,
      "Month": 12,
      "Year": 1996,
    },
    "dateRevised": {
      "Day": 1,
      "Month": 12,
      "Year": 2018,
    },
    "journalInfo": {
      "Country": "Hungary",
      "ISSNLinking": "0030-6002",
      "MedlineTA": "Orv Hetil",
      "NlmUniqueID": 376412,
    },
    "meshHeadings": [
      {
        "descriptorName": "Isotopes",
      },
      {
        "descriptorName": "Medicine",
      },
      {
        "descriptorName": "Radioisotopes",
      },
      {
        "descriptorName": "Research",
      },
    ],
  },
}
`;

exports[`syncPubmed Incremental Importation 1`] = `
{
  "_id": 17200418,
  "_seq": 4,
  "data": {
    "article": {
      "abstract": null,
      "authors": [
        {
          "firstName": "Youwen",
          "initials": "Y",
          "lastName": "Fang",
        },
        {
          "firstName": "Elaine",
          "initials": "E",
          "lastName": "Studer",
        },
        {
          "firstName": "Clint",
          "initials": "C",
          "lastName": "Mitchell",
        },
        {
          "firstName": "Steven",
          "initials": "S",
          "lastName": "Grant",
        },
        {
          "firstName": "William M",
          "initials": "WM",
          "lastName": "Pandak",
        },
        {
          "firstName": "Philip B",
          "initials": "PB",
          "lastName": "Hylemon",
        },
        {
          "firstName": "Paul",
          "initials": "P",
          "lastName": "Dent",
        },
      ],
      "journal": {
        "iSSN": "0026-895X",
        "isoAbbreviation": "Mol Pharmacol",
        "issue": 4,
        "pubDate": {
          "Month": "Apr",
          "Year": 2007,
        },
        "title": "Molecular pharmacology",
      },
      "language": "eng",
      "publicationTypeList": [
        "Journal Article",
        "Research Support, N.I.H., Extramural",
        "Research Support, Non-U.S. Gov't",
        "Research Support, U.S. Gov't, Non-P.H.S.",
      ],
      "title": "Conjugated bile acids regulate hepatocyte glycogen synthase activity in vitro and in vivo via Galphai signaling.",
    },
    "chemicals": [
      {
        "nameOfSubstance": "Bile Acids and Salts",
      },
      {
        "nameOfSubstance": "Deoxycholic Acid",
        "unii": "005990WHZZ",
      },
      {
        "nameOfSubstance": "Taurodeoxycholic Acid",
        "unii": "516-50-7",
      },
      {
        "nameOfSubstance": "Glycogen Synthase",
        "unii": "EC 2.4.1.11",
      },
      {
        "nameOfSubstance": "Phosphatidylinositol 3-Kinases",
        "unii": "EC 2.7.1.-",
      },
      {
        "nameOfSubstance": "Proto-Oncogene Proteins c-akt",
        "unii": "EC 2.7.11.1",
      },
      {
        "nameOfSubstance": "GTP-Binding Protein alpha Subunits, Gi-Go",
        "unii": "EC 3.6.5.1",
      },
    ],
    "dateCompleted": {
      "Day": 4,
      "Month": 6,
      "Year": 2007,
    },
    "dateRevised": {
      "Day": 1,
      "Month": 1,
      "Year": 2023,
    },
    "journalInfo": {
      "Country": "United States",
      "ISSNLinking": "0026-895X",
      "MedlineTA": "Mol Pharmacol",
      "NlmUniqueID": 35623,
    },
    "meshHeadings": [
      {
        "descriptorName": "Animals",
      },
      {
        "descriptorName": "Bile Acids and Salts",
        "qualifierName": "physiology",
      },
      {
        "descriptorName": "Cells, Cultured",
      },
      {
        "descriptorName": "Deoxycholic Acid",
        "qualifierName": "pharmacology",
      },
      {
        "descriptorName": "GTP-Binding Protein alpha Subunits, Gi-Go",
        "qualifierName": "physiology",
      },
      {
        "descriptorName": "Glycogen Synthase",
        "qualifierName": "metabolism",
      },
      {
        "descriptorName": "Hepatocytes",
      },
      {
        "descriptorName": "Male",
      },
      {
        "descriptorName": "Phosphatidylinositol 3-Kinases",
        "qualifierName": "metabolism",
      },
      {
        "descriptorName": "Proto-Oncogene Proteins c-akt",
        "qualifierName": "metabolism",
      },
      {
        "descriptorName": "Rats",
      },
      {
        "descriptorName": "Rats, Sprague-Dawley",
      },
      {
        "descriptorName": "Signal Transduction",
      },
      {
        "descriptorName": "Taurodeoxycholic Acid",
        "qualifierName": "pharmacology",
      },
    ],
  },
}
`;
